main/src/ca/uwaterloo/flix/language/phase/sjvm/GenLazyClasses.scala

```diff
@@ -0,0 +1,162 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase.sjvm
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.FinalAst.Root
+import ca.uwaterloo.flix.language.ast.PRefType._
+import ca.uwaterloo.flix.language.ast.PType._
+import ca.uwaterloo.flix.language.ast.RRefType._
+import ca.uwaterloo.flix.language.ast.RType._
+import ca.uwaterloo.flix.language.ast.{PType, RType}
+import ca.uwaterloo.flix.language.phase.sjvm.BytecodeCompiler._
+import ca.uwaterloo.flix.language.phase.sjvm.Instructions._
+
+/**
+ * Generates bytecode for the lazy classes.
+ */
+object GenLazyClasses {
+
+  val initializedFieldName: String = "initialized"
+  val initializedFieldType: RType[PInt32] = RBool()
+  val expressionFieldName: String = "expression"
+  val expressionFieldType: RReference[PAnyObject] = RReference(RObject())
+  val expressionToVoid: String = JvmName.objectToVoid
+  val valueFieldName: String = "value"
+  val forceMethod: String = "force"
+
+  /**
+   * Returns the set of lazy classes for the given set of types `ts`.
+   */
+  def gen()(implicit root: Root, flix: Flix): Map[JvmName, JvmClass] = {
+
+    // Generating each lazy class
+    def genAUX[T <: PType](valueFieldType: RType[T]): (JvmName, JvmClass) = {
+      val lazyType = RReference(RLazy(valueFieldType))
+      val bytecode = genByteCode(lazyType, valueFieldType)
+      lazyType.jvmName -> JvmClass(lazyType.jvmName, bytecode)
+    }
+
+    //Type that we need a cell class for
+    Map() +
+      genAUX(RBool()) +
+      genAUX(RInt8()) +
+      genAUX(RInt16()) +
+      genAUX(RInt32()) +
+      genAUX(RInt64()) +
+      genAUX(RChar()) +
+      genAUX(RFloat32()) +
+      genAUX(RFloat64()) +
+      genAUX(RReference(RObject()))
+  }
+
+  /**
+   * This method creates the class for each lazy value.
+   * The specific lazy class has an associated value type (tpe) which
+   * is either a jvm primitive or object.
+   *
+   * The lazy class has three fields - initialized: bool, expression: () -> tpe,
+   * and value: tpe. These are all private. force(context) is the only public
+   * method, which retuns a value of type tpe given a context to call the
+   * expression closure in.
+   *
+   * force will only evaluate the expression the first time, based on the flag initialized.
+   * After that point it will store the result in value and just return that.
+   */
+  private def genByteCode[T <: PType](lazyType: RReference[PLazy[T]], valueFieldType: RType[T])(implicit root: Root, flix: Flix): Array[Byte] = {
+    val className = lazyType.toInternalName
+    val classMaker = ClassMaker.openClassWriter(className, isFinal = true)
+
+    classMaker.makeField(initializedFieldName, initializedFieldType.toDescriptor, isStatic = false, isPublic = false)
+    classMaker.makeField(expressionFieldName, expressionFieldType.toDescriptor, isStatic = false, isPublic = false)
```

Comment:
I think it would be better to have:

`mkField`
`mkStaticField`

Reply:
Why mk instead of make? Consistency?